openhantek/src/widgets/dsowidget.h

```diff
@@ -24,10 +24,11 @@ struct ModelSpec;
 class GlScopeWindow;
 class DsoControl;
 
-class ChannelWidgets : protected QWidget {
+class ChannelWidgets : public QWidget {
```

Comment:
The QWidget class in Qt is very *feature rich* and overloaded.

Personally I'd like to hide the implementation details by making inherited QWidget **protected** and only relay certain required methods with the c++11 `using methodname = QWidget::methodname` syntax.

The reason is: The QtGui module has no future in Qt, QML is the preferred way to go. Let us be prepared by not making QWidget part of our interfaces.

Reply:
- DsoWidget must be capable of adding ChannelWidgets instance to a layout. This can be solved by having a method which returns you an instance of a widget (like root() you had before);
- It also seems to be impossible to use non-static methods as slots unless the class has a publicly accessible QObject base. This could be solved by passing lambdas to connect() instead of method pointers.

Do you really want these hacks just to have a protected QWidget base here? DsoWidget seems to have a public base.

Reply:
For connect to work it's not enough just to have the method public. The class instance must me castable to/from QObject. It comes from the implementation details of QtPrivate::QSlotObject template. impl() method takes QObject as an instance and tries to cast it to the method's declaring class. So, unless you have QObject as a public base, you can't use methods as slots, no matter public or not.